pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/TimeRangeSegmentPruner.java

```diff
@@ -0,0 +1,348 @@
+    }
+    List<Interval> ranges = getFilterTimeRange(filterQueryTree);
+    if (ranges == null) { // cannot prune based on time for input request
+      return segments;
+    }
+    if (ranges.size() == 0) { // invalid query time range
+      return Collections.emptySet();
+    }
+
+    Set<String> prunedSegments = new HashSet<>();
+    for (Interval range : ranges) {
+      Interval rangeMS = convertRangeToMS(range);
+      for (Object segment : _timeRangeMSToSegmentSearchTree.searchAll(rangeMS)) {
+        if (segments.contains(segment)) {
+          prunedSegments.add((String) segment);
+        }
+      }
+    }
+    return new HashSet<String>(prunedSegments);
+  }
+
+  private List<Interval> getFilterTimeRange(FilterQueryTree filterQueryTree) {
+    // return NUll if no time range info or cannot filter base on the info (e.g. 'SELECT * from myTable where time < 50 OR firstName = Jason')
+    // return an empty list if filtering range is specified but invalid (e.g. 'SELECT * from myTable where time < 50 AND time > 100')
+    switch (filterQueryTree.getOperator()) {
+      case AND:
+        List<List<Interval>> andRanges = new ArrayList<>();
+        for (FilterQueryTree child : filterQueryTree.getChildren()) {
+          List<Interval> childRanges = getFilterTimeRange(child);
+          if (childRanges != null) {
+            andRanges.add(childRanges);
+          }
+        }
+        return getIntersectionSortedRanges(andRanges);
+      case OR:
+        List<List<Interval>> orRanges = new ArrayList<>();
+        for (FilterQueryTree child : filterQueryTree.getChildren()) {
+          List<Interval> childRanges = getFilterTimeRange(child);
+          if (childRanges == null) {
+            return null;
+          } else {
+            orRanges.add(childRanges);
+          }
+        }
+        return getUnionSortedRanges(orRanges);
+      case EQUALITY:
+        if (filterQueryTree.getColumn().equals(_timeColumn)) {
+          long timeStamp = Long.parseLong(filterQueryTree.getValue().get(0));
+          return Collections.singletonList(new Interval(timeStamp, timeStamp));
+        }
+        return null;
+      case RANGE:
+        if (filterQueryTree.getColumn().equals(_timeColumn)) {
+          return parseRange(filterQueryTree.getValue());
+        }
+        return null;
+      default:
+        return null;
+    }
+  }
+
+  private Interval convertRangeToMS(Interval range) {
+    long min = range.min == MIN_START_TIME ? MIN_START_TIME : _timeUnit.toMillis(range.min);
+    long max = range.max == MAX_END_TIME ? MAX_END_TIME : _timeUnit.toMillis(range.max);
+    return new Interval(min, max);
+  }
+
+  private List<Interval> getIntersectionSortedRanges(List<List<Interval>> ranges) {
+    // Requires input ranges are sorted, the return ranges will be sorted
+    return getIntersectionSortedRanges(ranges, 0, ranges.size());
+  }
+
+  private List<Interval> getIntersectionSortedRanges(List<List<Interval>> ranges, int start, int end) {
```

Comment:
I think this and the next method can be simplified into one method that works on List<Interval> ranges